multipass-rke2.sh

```diff
@@ -1,160 +1,236 @@
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+# Setup microk8s for use within this code base.
+#
+# Requires: multipass
+#           jq
+#
+# Author(s): Sebastiaan van Steenis
+#            Justin Cook
+
+# shellcheck source=/dev/null
+. env.sh
+
+set -o errexit nounset
+
+## Configurable settings
 # Name for the cluster/configuration files
-NAME=""
-# Ubuntu image to use (xenial/bionic/focal)
-IMAGE="focal"
+NAME="generic"
+# Ubuntu image to use (xenial/bionic/focal/jammy)
+IMAGE="jammy"
 # RKE2 channel
-RKE2_CHANNEL="latest"
+RKE2_CHANNEL="stable"
 # RKE2 version
-#RKE2_VERSION="v1.25.7+rke2r1"
-# How many additional server machines to create
-SERVER_COUNT_MACHINE="0"
-# How many machines to create
-AGENT_COUNT_MACHINE="1"
+#RKE2_VERSION="v1.24.12+rke2r1"
+# How many master nodes to create
+MASTER_NODE_COUNT="1"
+# How many compute nodes to create
+AGENT_NODE_COUNT="2"
 # How many CPUs to allocate to each machine
-SERVER_CPU_MACHINE="2"
-AGENT_CPU_MACHINE="1"
-# How much disk space to allocate to each machine
-SERVER_DISK_MACHINE="10G"
-AGENT_DISK_MACHINE="10G"
+MASTER_NODE_CPU="2"
+AGENT_NODE_CPU="2"
+# How much disk space to allocate to each master and compute node
+MASTER_DISK_SIZE="20G"
+AGENT_DISK_SIZE="40G"
 # How much memory to allocate to each machine
-SERVER_MEMORY_MACHINE="4G"
-AGENT_MEMORY_MACHINE="2G"
+MASTER_MEMORY_SIZE="4G"
+AGENT_MEMORY_SIZE="8G"
 # Preconfigured secret to join the cluster (or autogenerated if empty)
+# Note: in order to use this script multiple times to add nodes, this needs
+# to be configured.
 TOKEN=""
-
-
-## Nothing to change after this line
-if [ -x "$(command -v multipass.exe)" > /dev/null 2>&1 ]; then
+# Hostnames or IPv4/IPv6 addresses as Subject Alternative Names on the server
+# TLS cert
+TLSSAN="rancher.test"
+## End configurable settings
+# Where to store the rke2 cluster kubeconfig
+LOCALKUBECONFIG="${HOME}/.kube/config-${NAME}"
+
+if [ -x "$(command -v multipass.exe)" ]
+then
     # Windows
     MULTIPASSCMD="multipass.exe"
-elif [ -x "$(command -v multipass)" > /dev/null 2>&1 ]; then
+elif [ -x "$(command -v multipass)" ]
+then
     # Linux/MacOS
     MULTIPASSCMD="multipass"
+    KUBECTL="$(command -v kubectl)"
 else
-    echo "The multipass binary (multipass or multipass.exe) is not available or not in your \$PATH"
+    echo "This platform dependencies are not available"
     exit 1
```

Comment:
I would keep the same log as it tells the user directly what is missing instead of it being more vague. Although multipass is pretty obvious as a requirement.

Reply:
And as `jq` is now a requirement, it needs a check for that as well otherwise users will only get notified after the process has already begun.

Reply:
Hi yes in my environment, env.sh checks requirements such as jq. Also given it's set errexit, the error will be in screen so extra output is redundant.

Reply:
As there is no `-x`, it won't print what it is happening afaik. If I run the script as-is, I just see;

```
$ ./multipass-rke2.sh 
This platform dependencies are not available
```

This still needs explaining to the user as they don't know what is required. I'm fine if this is just in the final echo before `exit 1`.

Reply:
Ok if a dep is not found in PATH it exits with specific output